sound/soc/sof/intel/hda-pcm.c

```diff
@@ -95,6 +95,78 @@ static inline u32 get_bits(struct snd_sof_dev *sdev, int sample_bits)
 	}
 };
 
+/* There are two dma modes in dsp:
+ * (1) host dma and link dma work in decouple mode for SOF + HDA
+ * (2) host dma and GP dma work together for SOF + I2S
+ * The difference is that in decopule mode, the host dma is set by dai
+ * frontend and link dma is set by dai backend and the setting maybe
+ * different while in I2S mode, host dma and GP dma are independent
+ * to each other and GP dma channel is allocated in FW instead of in host
+ */
+static inline int get_hda_dma_channel(struct snd_sof_dev *sdev,
+				      struct snd_pcm_substream *substream,
+				      struct sof_ipc_stream_params *ipc_params)
+{
+	struct snd_soc_pcm_runtime *fe = substream->private_data;
+	struct hdac_ext_stream *hda_stream = NULL;
+	struct snd_soc_dpcm *dpcm;
+	int direction = substream->stream;
+	struct sof_ipc_dai_hda_params *hda_params;
+
+	hda_params = &ipc_params->hda;
+	hda_params->hda_count = 0;
+
+	/* travel BE clients to get BE stream with link dma id */
+	list_for_each_entry(dpcm, &fe->dpcm[direction].be_clients, list_be) {
```

Comment:
Sorry, I really don't get this loop either.
What happens when you have a mix of Hdaudio and SSP/DMIC? Why would you loop on all BE clients, there should be something that checks if this is relevant?
Also as I mentioned it earlier in the review, don't you have cases where the information will be configured multiple times, or none?
I know we sound painful but essentially your design remains obscure, sorry.

Reply:
yes, actually there is a check in loop. Now I don't find a way to  check it is SSP or DMIC.

hda_stream = snd_soc_dai_get_dma_data(be->cpu_dai, substream);
if (hda_stream) {
...
here, hda_stream is only set by HDA dai backend. For other dai backend, it should be NULL.
And in FW, only HDA link dma use channel info.

For multiple dai backends, I need to check all the components of pipeline, form kernel, topology to Fw. We never test this case, so I need to spend some time to make it ready.

Reply:
I don't feel painful about implementing a feature. We have a plan with CCG team, so they often ask me which branch to use. This would trouble me a little.

Reply:
@RanderWang it would help if you describe the host registers used to get the channel info and the DSP link DMA registers you program the data into. That we we can read the docs and understand the problem better.